build_tools/cmake/iree_run_module_test.cmake

```diff
@@ -0,0 +1,222 @@
+#   DRIVER: Driver to run the module with. This can be overridden if it is
+#       provided by the benchmark_suite's flagfile.
+#   RUNNER_ARGS: additional args to pass to iree-run-module. The driver
+#       and input file are passed automatically.
+#   EXPECTED_OUTPUT: A file of expected output to compare with the output from
+#       iree-run-module
+#   LABELS: Additional labels to apply to the test. The package path and
+#       "driver=${DRIVER}" are added automatically.
+#   XFAIL: When set for x86_64|arm64-v8a|rv64, the test is expected to
+#       fail for the particular platforms/architectures for various reasons.
+#       e.g., benchmark_suite module is not available, upstream llvm backend
+#       breakage, etc.
+#   TIMEOUT: (optional) Test timeout.
+#
+# Example:
+#
+# iree_run_module_test(
+#   NAME
+#     person_detect_int8_correctness_test
+#   BENCHMARK_MODULE_SRC
+#     "TFLite/PersonDetect-int8"
+#   EXPECTED_OUTPUT
+#     "1x2xi8=[72 -72]"
+#   XFAIL
+#     "arm64-v8a"
+# )
+function(iree_run_module_test)
+  if(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME;BENCHMARK_MODULE_SRC;MODULE_SRC;DRIVER;EXPECTED_OUTPUT;TIMEOUT"
+    "RUNNER_ARGS;LABELS;XFAIL"
+    ${ARGN}
+  )
+
+  if(CMAKE_CROSSCOMPILING AND "hostonly" IN_LIST _RULE_LABELS)
+    return()
+  endif()
+
+  if(_RULE_BENCHMARK_MODULE_SRC AND _RULE_MODULE_SRC)
+    message(SEND_ERROR "Can only specify either MODULE_SRC or BENCHMARK_MODULE_SRC")
+  endif()
+
+  if(NOT _RULE_BENCHMARK_MODULE_SRC AND NOT _RULE_MODULE_SRC)
+    message(SEND_ERROR "Need to specify either MODULE_SRC or BENCHMARK_MODULE_SRC")
+  endif()
+
+  # Benchmark suite needs to be stored at the location of
+  # `IREE_BENCHMARK_SUITE_DIR` or the test target is bypassed.
+  if(_RULE_BENCHMARK_MODULE_SRC AND NOT DEFINED IREE_BENCHMARK_SUITE_DIR)
+    return()
+  endif()
+
+  set(_RUNNER_TARGET "iree-run-module")
+  iree_package_name(_PACKAGE_NAME)
+  set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+  if(_RULE_XFAIL)
+    platform_xfail(_XFAIL ${_RULE_XFAIL})
+  endif()
+
+  if(_RULE_MODULE_SRC)
+    get_filename_component(_SRC "${_RULE_MODULE_SRC}" REALPATH)
+  elseif(_RULE_BENCHMARK_MODULE_SRC)
+    set(_MODULE_FLAG_DIR "${IREE_BENCHMARK_SUITE_DIR}/${_RULE_BENCHMARK_MODULE_SRC}/")
+    # Find the platform specific module flag file with matching path name.
+    # TODO(#10391): Update this logic with the new benchmark framework.
+    if(CMAKE_SYSTEM_PROCESSOR STREQUAL "riscv64" AND
+       CMAKE_SYSTEM_NAME STREQUAL "Linux")
+      set(_FLAGFILE_HINT_PATH "${_MODULE_FLAG_DIR}/iree-llvm-cpu*RV64*__full-inference,default-flags/flagfile")
+    elseif(ANDROID)
+      set(_FLAGFILE_HINT_PATH "${_MODULE_FLAG_DIR}/iree-llvm-cpu*ARM64-v8A*__big-core,full-inference,default-flags/flagfile")
+    else()  # X86_64
+      set(_FLAGFILE_HINT_PATH "${_MODULE_FLAG_DIR}/iree-llvm-cpu*x86_64*__full-inference,default-flags/flagfile")
+    endif()
```

Comment:
Oof, this is using some very specific implementation details of other build system functions. Can these be explicitly passed through, rather than be inferred like this?

This looks like inverted layering - tests should not be aware of benchmarks.

Maybe teach the benchmarking suite functions a way to define both a test (using this) and a benchmark (using what they already have), and pass through the flags/paths at that higher level?

Reply:
I think the issue here is partially that the benchmarks really aren't limited to being benchmarks. They're just configurations for compiling and running a model. @pzread is fixing this over-specific naming in his new work, which is why I didn't push back too hard. I agree the flag file parsing is gross. If these were not called "benchmark" though, how concerned would you be?

Reply:
> If these were not called "benchmark" though, how concerned would you be?

Still pretty concerned. I'd want to see this function to be a (relative) simple building block - given a compiled program, a set of flags, and the expected output, run the program through `iree-run-module` with those flags and check that the output matches.

Here, a different function is generating a bunch of files - that function should be passing what it generates in to this function, rather than having this function need to understand the structure produced over there.

---

I should be able to write a (for example) "pytorch correctness test" using this function without needing to change the function:
* Import program from Pytorch, snapshot expected output
* Compile program using `iree-compile`
* Add `iree-run-module`test with that expected output

If my import/compile steps there use some other directory structure, I shouldn't need to make changes here.

---

The platform specific stuff (any branches on Android/RISC-V/etc.) here is spreading in some weird ways too. I'm wondering if we can introduce some helper functions to organize that logic.

Reply:
I believe these suggestions are all part of the rewrite in https://github.com/iree-org/iree/pull/10391. The logic here is to match the benchmark structure so the test can find the right vmfb and flagfile. Jerry and I had the discussion in https://github.com/iree-org/iree/pull/10391#issuecomment-1256574523

Reply:
I'd much rather we start with a simple `iree_run_module_test()` that `iree_benchmark_*()` uses, instead of entangling `iree_run_module_test()` with _anything_ benchmarking-related from the very start.

Reply:
Untangled the two functions in the new commit.

Reply:
Thanks. I'm liking this more now, but it's still difficult to review this PR when it does so many things at once. I'd rather we start with a PR that just adds the new type of test (with at least one usage in https://github.com/iree-org/iree/tree/main/tools/test), then follow up with a PR that connects the programs in the benchmarks suite to the new rule and another PR that builds/runs those tests on the CI.

I'm concerned that this new building block test is getting entangled in ongoing benchmark infrastructure refactoring and I want to make sure that it is set up to be as stable as `iree_bytecode_module`, `iree_check_test`, and the other build/test functions that have persisted through a few rounds of similar refactorings.

Reply:
I will refactor the PR into three (The original PR has split into several commits with the same intent, but sorry I didn't mark it clearly in the PR message)  1) basic test 2) benchmark suite CI, 3) x86_64 benchmark suite CI.

Reply:
Thanks. When a PR gets up to 50+ comments (we're at 77 now), that's a pretty good sign that splitting into a few incremental PRs would be easier to review - and hopefully easier to author too :)